feat: nvidia-persistenced to Nvidia kmod packages

isaac-400 opened the pull request:
Closes bottlerocket-os/bottlerocket#3960

**Description of changes:**

Whenever the NVIDIA device resources are no longer in use, the NVIDIA
kernel driver will tear down the device state. `nvidia-persistenced`
activates persistence mode, which keeps the device files open preventing
the kernel from removing the device state [[1]]. This is desirable in
applications that may suffer performance hits due to repeated device
initialization.

The NVIDIA device drivers ship with templates for running
`nvidia-persistenced` as a systemd unit [[2]]. This change uses that template.

The `nvidia-persistenced` documentation advises that while the systemd
unit can run as root, the unit should provide a non-root user under which
`nvidia-persistenced` will run. This change adds a non-root user for `nvidia-persistenced`. 

See the documentation included with the NVIDIA driver for more
information about `nvidia-persistenced` [[3]].

[1]: https://docs.nvidia.com/deploy/driver-persistence/index.html
[2]: https://github.com/NVIDIA/nvidia-persistenced/blob/main/init/systemd/nvidia-persistenced.service.template
[3]: https://github.com/NVIDIA/nvidia-persistenced

**Testing Done:**

ECS: The unit sets the devices to persistence mode, and we can see that the
device is set correctly inside a task container.

k8s: Node successfully joined cluster and persistence mode is enabled inside test pod.

**aws-ecs-1-nvidia**

``` example
bash-5.1# apiclient get os
{
  "os": {
    "arch": "x86_64",
    "build_id": "39f91d5d-dirty",
    "pretty_name": "Bottlerocket OS 1.21.1 (aws-ecs-1-nvidia)",
    "variant_id": "aws-ecs-1-nvidia",
    "version_id": "1.21.1"
  }
}
bash-5.1# docker ps
CONTAINER ID   IMAGE                                  COMMAND                  CREATED          STATUS          PORTS     NAMES
b90b0fec4e7e   fedora                                 "sh -c 'sleep infini…"   38 seconds ago   Up 37 seconds             ecs-ecs-test-4-main-948ee88ccac2abbd6d00
6a16e877533f   amazon/amazon-ecs-pause:bottlerocket   "/usr/bin/pause"         44 seconds ago   Up 44 seconds             ecs-ecs-test-4-internalecspause-f4fbfab2fde3e3f60600
bash-5.1# docker exec -it b90b0fec4e7e nvidia-smi
Tue Sep  3 19:03:36 2024
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.256.02   Driver Version: 470.256.02   CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A10G         On   | 00000000:00:1E.0 Off |                    0 |
|  0%   28C    P8     9W / 300W |      0MiB / 22731MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
bash-5.1# systemctl status nvidia-persistenced
● nvidia-persistenced.service - NVIDIA Persistence Daemon
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvidia-persistenced.service; enabled; preset: enabled)
     Active: active (running) since Tue 2024-09-03 19:00:29 UTC; 3min 17s ago
   Main PID: 1914 (nvidia-persiste)
      Tasks: 1 (limit: 18928)
     Memory: 4.5M
     CGroup: /system.slice/nvidia-persistenced.service
             └─1914 /x86_64-bottlerocket-linux-gnu/sys-root/usr/libexec/nvidia/tesla/bin/470.256.02/nvidia-persistenced --user nvidia

Sep 03 19:00:28 ip-10-194-20-18.us-west-2.compute.internal nvidia-smi[1889]:
Sep 03 19:00:28 ip-10-194-20-18.us-west-2.compute.internal nvidia-smi[1889]: +-----------------------------------------------------------------------------+
Sep 03 19:00:28 ip-10-194-20-18.us-west-2.compute.internal nvidia-smi[1889]: | Processes:                                                                  |
Sep 03 19:00:28 ip-10-194-20-18.us-west-2.compute.internal nvidia-smi[1889]: |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
Sep 03 19:00:28 ip-10-194-20-18.us-west-2.compute.internal nvidia-smi[1889]: |        ID   ID                                                   Usage      |
Sep 03 19:00:28 ip-10-194-20-18.us-west-2.compute.internal nvidia-smi[1889]: |=============================================================================|
Sep 03 19:00:28 ip-10-194-20-18.us-west-2.compute.internal nvidia-smi[1889]: |  No running processes found                                                 |
Sep 03 19:00:28 ip-10-194-20-18.us-west-2.compute.internal nvidia-smi[1889]: +-----------------------------------------------------------------------------+
Sep 03 19:00:28 ip-10-194-20-18.us-west-2.compute.internal nvidia-persistenced[1914]: Started (1914)
Sep 03 19:00:29 ip-10-194-20-18.us-west-2.compute.internal systemd[1]: Started NVIDIA Persistence Daemon.
bash-5.1#
```

**aws-ecs-2-nvidia**

``` example
bash-5.1# apiclient get os
{
  "os": {
    "arch": "x86_64",
    "build_id": "39f91d5d-dirty",
    "pretty_name": "Bottlerocket OS 1.21.1 (aws-ecs-2-nvidia)",
    "variant_id": "aws-ecs-2-nvidia",
    "version_id": "1.21.1"
  }
}
bash-5.1# docker ps
CONTAINER ID   IMAGE                                  COMMAND                  CREATED          STATUS          PORTS     NAMES
e87a4e93439c   fedora                                 "sh -c 'sleep infini…"   8 seconds ago    Up 6 seconds              ecs-ecs-test-2-main-8883a48ba4b884efda01
3eca5efed891   amazon/amazon-ecs-pause:bottlerocket   "/usr/bin/pause"         13 seconds ago   Up 12 seconds             ecs-ecs-test-2-internalecspause-96db8af6f891edf5b501
bash-5.1# docker exec -it e87a4e93439c nvidia-smi
Wed Sep  4 16:57:12 2024
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A10G                    On  | 00000000:00:1E.0 Off |                    0 |
|  0%   26C    P8               8W / 300W |      0MiB / 23028MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+
bash-5.1# systemctl status nvidia-persistenced
● nvidia-persistenced.service - NVIDIA Persistence Daemon
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvidia-persistenced.service; enabled; preset: enabled)
     Active: active (running) since Wed 2024-09-04 16:42:22 UTC; 15min ago
   Main PID: 1430 (nvidia-persiste)
      Tasks: 1 (limit: 18906)
     Memory: 41.3M
        CPU: 4.445s
     CGroup: /system.slice/nvidia-persistenced.service
             └─1430 /x86_64-bottlerocket-linux-gnu/sys-root/usr/libexec/nvidia/tesla/bin/nvidia-persistenced --user nvidia

Sep 04 16:42:20 ip-172-31-83-142.ec2.internal nvidia-smi[1412]:
Sep 04 16:42:20 ip-172-31-83-142.ec2.internal nvidia-smi[1412]: +---------------------------------------------------------------------------------------+
Sep 04 16:42:20 ip-172-31-83-142.ec2.internal nvidia-smi[1412]: | Processes:                                                                            |
Sep 04 16:42:20 ip-172-31-83-142.ec2.internal nvidia-smi[1412]: |  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
Sep 04 16:42:20 ip-172-31-83-142.ec2.internal nvidia-smi[1412]: |        ID   ID                                                             Usage      |
Sep 04 16:42:20 ip-172-31-83-142.ec2.internal nvidia-smi[1412]: |=======================================================================================|
Sep 04 16:42:20 ip-172-31-83-142.ec2.internal nvidia-smi[1412]: |  No running processes found                                                           |
Sep 04 16:42:20 ip-172-31-83-142.ec2.internal nvidia-smi[1412]: +---------------------------------------------------------------------------------------+
Sep 04 16:42:20 ip-172-31-83-142.ec2.internal nvidia-persistenced[1430]: Started (1430)
Sep 04 16:42:22 ip-172-31-83-142.ec2.internal systemd[1]: Started NVIDIA Persistence Daemon.
bash-5.1#
```

**aws-k8s-1.23-nvidia**

``` example
[nix-shell:~/workplace/bottlerocket]$  kubectl exec gpu-test -c main -- nvidia-smi
 kubectl exec gpu-test -c main -- nvidia-smi
Tue Sep  3 23:10:20 2024       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.256.02   Driver Version: 470.256.02   CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A10G         On   | 00000000:00:1E.0 Off |                    0 |
|  0%   25C    P8     9W / 300W |      0MiB / 22731MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

[nix-shell:~/workplace/bottlerocket]$  kubectl exec gpu-test -c main -- nvidia-smi --query-gpu=gpu_name,persistence_mode --format=csv
 kubectl exec gpu-test -c main -- nvidia-smi --query-gpu=gpu_name,persistence_mode --format=csv
name, persistence_mode
NVIDIA A10G, Enabled
```

**aws-k8s-1.26-nvidia**

``` example
[nix-shell:~/workplace/bottlerocket]$  kubectl exec gpu-test -c main -- nvidia-smi
 kubectl exec gpu-test -c main -- nvidia-smi
Wed Sep  4 16:59:43 2024
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A10G                    On  | 00000000:00:1E.0 Off |                    0 |
|  0%   24C    P8               9W / 300W |      0MiB / 23028MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+


[nix-shell:~/workplace/bottlerocket]$  kubectl exec gpu-test -c main -- nvidia-smi --query-gpu=gpu_name,persistence_mode --format=csv
 kubectl exec gpu-test -c main -- nvidia-smi --query-gpu=gpu_name,persistence_mode --format=csv
name, persistence_mode
NVIDIA A10G, Enabled
```

**aws-k8s-1.30-nvidia**

``` example
[nix-shell:~/workplace/bottlerocket]$  kubectl exec gpu-test -c main -- nvidia-smi
 kubectl exec gpu-test -c main -- nvidia-smi
Wed Sep  4 17:25:29 2024
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A10G                    On  | 00000000:00:1E.0 Off |                    0 |
|  0%   26C    P8              12W / 300W |      0MiB / 23028MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+

[nix-shell:~/workplace/bottlerocket]$  kubectl exec gpu-test -c main -- nvidia-smi --query-gpu=gpu_name,persistence_mode --format=csv
 kubectl exec gpu-test -c main -- nvidia-smi --query-gpu=gpu_name,persistence_mode --format=csv
name, persistence_mode
NVIDIA A10G, Enabled
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is
dual-licensed under the terms of both the Apache License, version 2.0,
and the MIT license.

Commit:
kmod-5.10-nvidia: add nvidia-persistenced

Whenever the NVIDIA device resources are no longer in use, the NVIDIA
kernel driver will tear down the device state. `nvidia-persistenced`
activates persistence mode, which keeps the device files open which
prevents the kernel from removing the device state. This is desirable
in applications that may suffer performance hits due to repeated
device initialization.

The NVIDIA device drivers ship with templates for running
`nvidia-persistenced` as a systemd unit. This change uses that
template.

The `nvidia-persistenced` documentation advises that while the systemd
unit can run as root, the unit should provide a non-root user for
`nvidia-persistenced` to run under.

See the documentation included with the NVIDIA driver for more
information about `nvidia-persistenced`.

Commit:
kmod-5.15-nvidia: add nvidia-persistenced

Commit:
kmod-6.1-nvidia: add nvidia-persistenced

arnaldo2792 commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -20,6 +20,9 @@ Source2: NVidiaEULAforAWS.pdf
 Source202: nvidia-dependencies-modules-load.conf
 
+Source210: nvidia-sysusers.conf
+Source211: nvidia-persistenced.service.in
```
You could continue with the existing sequence, since these are common files.

arnaldo2792 commented on packages/kmod-5.10-nvidia/nvidia-persistenced.service.in:
```diff
@@ -0,0 +1,40 @@
+Type=forking
+# Run the NVIDIA System Management Interface to create the device nodes.
+ExecStartPre=__NVIDIA_BINDIR__/nvidia-smi
+ExecStart=__NVIDIA_BINDIR__/nvidia-persistenced --user __USER__
```
You know upfront what the user is, so you could just hard code the value.

arnaldo2792 commented on packages/kmod-5.10-nvidia/nvidia-persistenced.service.in:
```diff
@@ -0,0 +1,40 @@
+#
+# This is a sample systemd service file, designed to show how the NVIDIA
+# Persistence Daemon can be started.
+#
```
This isn't needed, the service is different enough from the service provided by NVIDIA.

arnaldo2792 commented on packages/kmod-5.10-nvidia/nvidia-persistenced.service.in:
```diff
@@ -0,0 +1,40 @@
+After=load-kernel-modules.service
+
+[Service]
+Type=forking
```
Systemd discourages the use of `forking`:

> If set to forking, the manager will consider the unit started immediately after the binary that forked off by the manager exits. The use of this type is discouraged, use notify, notify-reload, or dbus instead.

Since this is a daemon, the default `simple` should be enough

isaac-400 replied:
I ran some tests when I set the mode to simple, it looks like the unit immediately exits. 

```
○ nvidia-persistenced.service - NVIDIA Persistence Daemon
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvidia-persistenced.service; enabled; preset: enabled)
     Active: inactive (dead) since Fri 2024-09-06 17:24:31 UTC; 1min 50s ago
   Duration: 2.102s
   Main PID: 1424 (code=exited, status=0/SUCCESS)
        CPU: 2.814s

Sep 06 17:24:28 ip-172-31-92-66.ec2.internal nvidia-smi[1407]: |  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
Sep 06 17:24:28 ip-172-31-92-66.ec2.internal nvidia-smi[1407]: |        ID   ID                                                             Usage      |
Sep 06 17:24:28 ip-172-31-92-66.ec2.internal nvidia-smi[1407]: |=======================================================================================|
Sep 06 17:24:28 ip-172-31-92-66.ec2.internal nvidia-smi[1407]: |  No running processes found                                                           |
Sep 06 17:24:28 ip-172-31-92-66.ec2.internal nvidia-smi[1407]: +---------------------------------------------------------------------------------------+
Sep 06 17:24:29 ip-172-31-92-66.ec2.internal systemd[1]: Started NVIDIA Persistence Daemon.
Sep 06 17:24:29 ip-172-31-92-66.ec2.internal nvidia-persistenced[1426]: Started (1426)
Sep 06 17:24:31 ip-172-31-92-66.ec2.internal nvidia-persistenced[1426]: Shutdown (1426)
Sep 06 17:24:31 ip-172-31-92-66.ec2.internal systemd[1]: nvidia-persistenced.service: Deactivated successfully.
Sep 06 17:24:31 ip-172-31-92-66.ec2.internal systemd[1]: nvidia-persistenced.service: Consumed 2.814s CPU time.
bash-5.1# ps axu | grep persistenced
root        3274  0.0  0.0   3056  1508 pts/1    S+   17:55   0:00 grep persistenced
```

It seems like it only spawns the daemon when I run `nvidia-persistenced` from the command line.

```
bash-5.1# ./x86_64-bottlerocket-linux-gnu/sys-root/usr/libexec/nvidia/tesla/bin/nvidia-persistenced
bash-5.1# ps axu | grep persistenced
root        3279 48.7  0.0   4972  1660 ?        Ss   17:56   0:02 ./x86_64-bottlerocket-linux-gnu/sys-root/usr/libexec/nvidia/tesla/bin/nvidia-persistenced
root        3289  0.0  0.0   3056  1476 pts/1    S+   17:56   0:00 grep persistenced
```

I don't think simple is the correct unit type here. Maybe `notify`? Nvidia's provided template chooses `forking`.

arnaldo2792 replied:
I read the `nvidia-persistenced` code, and the main parent process forks the daemon in a child process. This comment was helpful to understand the behavior:

https://github.com/NVIDIA/nvidia-persistenced/blob/39e71b2cb7542d14f5106043fb4c7a789428144b/nvidia-persistenced.c#L220

`forking` seems to be the only choice we have since `nvidia-persistenced` doesn't integrate with `sd_notify`.

arnaldo2792 commented on packages/kmod-5.10-nvidia/nvidia-persistenced.service.in:
```diff
@@ -0,0 +1,40 @@
+[Service]
+Type=forking
+# Run the NVIDIA System Management Interface to create the device nodes.
+ExecStartPre=__NVIDIA_BINDIR__/nvidia-smi
```
According to the `nvidia-persistenced` man pages:

> The daemon indirectly utilizes nvidia-modprobe via the nvidia-cfg library to load the NVIDIA kernel module and create the NVIDIA character device files after the daemon has dropped its root privileges, if it had any to begin with. If nvidia-modprobe is not installed, the daemon may not be able to start properly if it is not run with root privileges. 

So I assume you called `nvidia-smi` to circumvent the problem of creating the devices. Do you have the same problem if you run `nvidia-persistenced` as UID 0?

isaac-400 replied:
When I run `nvidia-persistenced` with UID 0 it seems to create the device files just fine. It's only when I provide a user that it struggles. I'm guessing having the daemon run under a user is a better security practice, so that's why I've gone to some lengths to do the `--user` approach.

arnaldo2792 replied:
`nvidia-smi` calls are a bit expensive. It won't only create the devices, but  it actually queries them so this operation may slowdown the boot time.

Regarding changing the user of the daemon, the man pages mention that `nvidia-modprobe` will be executed to create the devices but this binary has to be installed as `setuid` and at `/usr/bin/` :

> The application nvidia-modprobe ('/usr/bin/nvidia-modprobe') is installed as setuid root and is used to load the NVIDIA kernel module and create the '/dev/nvidia*' device nodes by processes that don't run with sufficient privileges to do those things themselves.

So even if we provide `nvidia-modprobe` (which is already provided in the run archive, and we could include it), the devices are still created by UID 0 (the privilege we were trying to drop).

I think that changing the UID of the daemon doesn't really get us too much, and it should be OK to just run the daemon as UID 0. Changing the UID will require us to include one more binary which doesn't seem necessary when running as UID 0, since the binary itself requires to be `setuid` to be able to create the devices. And this binary isn't even used today in containers, so we don't get too much value out of it by adding it to the distro.

If we are concerned about the privileges of the daemon, we could use systemd directives to restrict its privileges, e.g.:

```
NoNewPrivileges=true
```

https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#NoNewPrivileges=

You could even play around with:

https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#SystemCallFilter=

However, we don't know what system calls the daemon (or any of the libraries it depends on) run, it will be rather hard to nail down the correct filter.

arnaldo2792 reviewed (commented):
Changes look good! Thanks for the contribution!

arnaldo2792 commented on packages/kmod-5.10-nvidia/kmod-5.10-nvidia.spec:
```diff
@@ -148,10 +149,15 @@ install -m 755 nvidia-smi %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{te
 install -m 755 nvidia-debugdump %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{tesla_470}
 install -m 755 nvidia-cuda-mps-control %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{tesla_470}
 install -m 755 nvidia-cuda-mps-server %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{tesla_470}
+install -m 755 nvidia-persistenced %{buildroot}%{_cross_libexecdir}/nvidia/tesla/bin/%{tesla_470}
```
I thought about installing `nvidia-persistenced` in `%{_cross_bindir}`, since it is meant to be a system service. The problem with this approach is that if in the future NVIDIA ships other `run` archives different than the `tesla` archive and we have to include it, there might be two versions of `nvidia-persistenced` that would have to be shipped.

I think we can keep it as it is, and going forward, we could do some guessing at runtime based on the driver that was loaded to override the path to `nvidia-persistenced` using systemd drop-ins.

arnaldo2792 commented on packages/kmod-5.15-nvidia/nvidia-persistenced.service.in:
```diff
@@ -0,0 +1,14 @@
+[Service]
+Type=forking
+# Run the NVIDIA System Management Interface to create the device nodes.
+ExecStart=__NVIDIA_BINDIR__/nvidia-persistenced
```
In 5.15 and 6.1, the path to `nvidia-persistenced` is more consistent and deterministic (it doesn't include the version). It should be fine to just use the full path here and you will save yourself the extra `sed`.

isaac-400 commented:
Tested the latest round of changes on `aws-ecs-2-nvidia`:

```
bash-5.1# apiclient get os
{
  "os": {
    "arch": "x86_64",
    "build_id": "a63007cf-dirty",
    "pretty_name": "Bottlerocket OS 1.22.0 (aws-ecs-2-nvidia)",
    "variant_id": "aws-ecs-2-nvidia",
    "version_id": "1.22.0"
  }
}
bash-5.1# docker ps
CONTAINER ID   IMAGE                                         COMMAND                  CREATED          STATUS          PORTS     NAMES
8485910eb8c0   public.ecr.aws/docker/library/fedora:latest   "sh -c 'sleep infini…"   15 seconds ago   Up 13 seconds             ecs-gpu-test-24-main-e6be94f7b6d2ccac1d00
77960735e8cd   amazon/amazon-ecs-pause:bottlerocket          "/usr/bin/pause"         20 seconds ago   Up 19 seconds             ecs-gpu-test-24-internalecspause-9eebfbbcf3f4f2c03300
bash-5.1# docker exec -it 8485910eb8c0 nvidia-smi
Thu Sep 12 17:02:56 2024
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A10G                    On  | 00000000:00:1E.0 Off |                    0 |
|  0%   23C    P8               9W / 300W |      0MiB / 23028MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+
bash-5.1# systemctl status nvidia-persistenced
● nvidia-persistenced.service - NVIDIA Persistence Daemon
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvidia-persistenced.service; enabled; preset: enabled)
     Active: active (running) since Thu 2024-09-12 16:58:18 UTC; 5min ago
   Main PID: 1415 (nvidia-persiste)
      Tasks: 1 (limit: 18906)
     Memory: 37.0M
        CPU: 1.949s
     CGroup: /system.slice/nvidia-persistenced.service
             └─1415 /usr/libexec/nvidia/tesla/bin/nvidia-persistenced

Sep 12 16:58:16 ip-10-194-20-100.us-west-2.compute.internal systemd[1]: Starting NVIDIA Persistence Daemon...
Sep 12 16:58:16 ip-10-194-20-100.us-west-2.compute.internal nvidia-persistenced[1415]: Started (1415)
Sep 12 16:58:18 ip-10-194-20-100.us-west-2.compute.internal systemd[1]: Started NVIDIA Persistence Daemon.
bash-5.1#
```

bcressey commented on packages/kmod-6.1-nvidia/nvidia-persistenced.service:
```diff
@@ -0,0 +1,14 @@
+
+[Service]
+Type=forking
+# Run the NVIDIA System Management Interface to create the device nodes.
```
This comment doesn't seem to apply?

bcressey commented on packages/kmod-6.1-nvidia/nvidia-persistenced.service:
```diff
@@ -0,0 +1,14 @@
+Type=forking
+# Run the NVIDIA System Management Interface to create the device nodes.
+ExecStart=/usr/libexec/nvidia/tesla/bin/nvidia-persistenced
+ExecStopPost=/bin/rm -rf /var/run/nvidia-persistenced
```
Would be good to clean up before as well, in case the service doesn't shut down cleanly:
```suggestion
ExecStartPre=/bin/rm -rf /run/nvidia-persistenced
ExecStart=/usr/libexec/nvidia/tesla/bin/nvidia-persistenced
ExecStopPost=/bin/rm -rf /run/nvidia-persistenced
```

bcressey commented:
This [older doc](https://download.nvidia.com/XFree86/Linux-x86_64/396.51/README/nvidia-persistenced.html) says:
> It is strongly recommended, though not required, that the daemon be run as a non-root user for security purposes.

So I really would like to bring back the `--user` option, or run the daemon directly as that user via `User=`.

bcressey commented:
I'd recommend including `nvidia-modprobe` in the package, but not installing it setuid. Then you could add a second systemd unit that invokes it as root, which `nvidia-persistenced.service` would depend on.

Commit:
kmod-6.1-nvidia: add nvidia-modprobe

Commit:
kmod-5.15-nvidia: add nvidia-modprobe

Commit:
kmod-5.10-nvidia: add nvidia-modprobe

isaac-400 commented:
### Summary of changes
- Install `nvidia-modprobe` as setuid.
- Use the `--user` option with a non-root user, so `nvidia-persistenced` can drop root permissions.
- Removed the `NoNewPermissions` systemd option, since this inhibits the non-root user from calling `nvidia-modprobe` setuid.
- Modified the systemd units to be compatible with recent `load-kernel-modules` changes.


### Testing

#### aws-ecs-1-nvidia
```
bash-5.1# apiclient get os
{
  "os": {
    "arch": "x86_64",
    "build_id": "26ec2cc5-dirty",
    "pretty_name": "Bottlerocket OS 1.23.0 (aws-ecs-1-nvidia)",
    "variant_id": "aws-ecs-1-nvidia",
    "version_id": "1.23.0"
  }
}
bash-5.1# systemctl status nvidia-persistenced
● nvidia-persistenced.service - NVIDIA Persistence Daemon
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvidia-persistenced.service; enabled; preset: enabled)
     Active: active (running) since Fri 2024-10-04 15:58:20 UTC; 1min 29s ago
   Main PID: 1942 (nvidia-persiste)
      Tasks: 1 (limit: 18928)
     Memory: 576.0K
     CGroup: /system.slice/nvidia-persistenced.service
             └─1942 /x86_64-bottlerocket-linux-gnu/sys-root/usr/libexec/nvidia/tesla/bin/470.256.02/nvidia-persistenced --user nvidia --verbose

Oct 04 15:58:19 ip-10-194-20-226.us-west-2.compute.internal systemd[1]: Starting NVIDIA Persistence Daemon...
Oct 04 15:58:19 ip-10-194-20-226.us-west-2.compute.internal nvidia-persistenced[1942]: Verbose syslog connection opened
Oct 04 15:58:19 ip-10-194-20-226.us-west-2.compute.internal nvidia-persistenced[1942]: Now running with user ID 981 and group ID 981
Oct 04 15:58:19 ip-10-194-20-226.us-west-2.compute.internal nvidia-persistenced[1942]: Started (1942)
Oct 04 15:58:19 ip-10-194-20-226.us-west-2.compute.internal nvidia-persistenced[1942]: device 0000:00:1e.0 - registered
Oct 04 15:58:20 ip-10-194-20-226.us-west-2.compute.internal nvidia-persistenced[1942]: device 0000:00:1e.0 - persistence mode enabled.
Oct 04 15:58:20 ip-10-194-20-226.us-west-2.compute.internal nvidia-persistenced[1942]: device 0000:00:1e.0 - NUMA memory onlined.
Oct 04 15:58:20 ip-10-194-20-226.us-west-2.compute.internal nvidia-persistenced[1942]: Local RPC services initialized
Oct 04 15:58:20 ip-10-194-20-226.us-west-2.compute.internal systemd[1]: Started NVIDIA Persistence Daemon.
bash-5.1#
```

#### aws-k8s-1.24-nvidia
```
bash-5.1# apiclient get os
{
  "os": {
    "arch": "x86_64",
    "build_id": "26ec2cc5-dirty",
    "pretty_name": "Bottlerocket OS 1.23.0 (aws-k8s-1.24-nvidia)",
    "variant_id": "aws-k8s-1.24-nvidia",
    "version_id": "1.23.0"
  }
}
bash-5.1#
bash-5.1# systemctl status nvidia-persistenced
● nvidia-persistenced.service - NVIDIA Persistence Daemon
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvidia-persistenced.service; enabled; preset: enabled)
     Active: active (running) since Fri 2024-10-04 16:31:49 UTC; 32min ago
   Main PID: 1365 (nvidia-persiste)
      Tasks: 1 (limit: 18929)
     Memory: 37.1M
     CGroup: /system.slice/nvidia-persistenced.service
             └─1365 /usr/libexec/nvidia/tesla/bin/nvidia-persistenced --user nvidia --verbose

Oct 04 16:31:43 ip-192-168-19-182.ec2.internal systemd[1]: Starting NVIDIA Persistence Daemon...
Oct 04 16:31:43 ip-192-168-19-182.ec2.internal nvidia-persistenced[1365]: Verbose syslog connection opened
Oct 04 16:31:43 ip-192-168-19-182.ec2.internal nvidia-persistenced[1365]: Now running with user ID 981 and group ID 981
Oct 04 16:31:43 ip-192-168-19-182.ec2.internal nvidia-persistenced[1365]: Started (1365)
Oct 04 16:31:43 ip-192-168-19-182.ec2.internal nvidia-persistenced[1365]: device 0000:00:1e.0 - registered
Oct 04 16:31:49 ip-192-168-19-182.ec2.internal nvidia-persistenced[1365]: device 0000:00:1e.0 - persistence mode enabled.
Oct 04 16:31:49 ip-192-168-19-182.ec2.internal nvidia-persistenced[1365]: device 0000:00:1e.0 - NUMA memory onlined.
Oct 04 16:31:49 ip-192-168-19-182.ec2.internal nvidia-persistenced[1365]: Local RPC services initialized
Oct 04 16:31:49 ip-192-168-19-182.ec2.internal systemd[1]: Started NVIDIA Persistence Daemon.
bash-5.1#
bash-5.1# uname -a
Linux ip-192-168-19-182.ec2.internal 5.15.167 #1 SMP Fri Oct 4 02:55:35 UTC 2024 x86_64 GNU/Linux
```

### aws-ecs-2-nvidia
```
bash-5.1# apiclient get os
{
  "os": {
    "arch": "x86_64",
    "build_id": "26ec2cc5-dirty",
    "pretty_name": "Bottlerocket OS 1.23.0 (aws-ecs-2-nvidia)",
    "variant_id": "aws-ecs-2-nvidia",
    "version_id": "1.23.0"
  }
}
bash-5.1# systemctl status nvidia-persistenced
● nvidia-persistenced.service - NVIDIA Persistence Daemon
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvidia-persistenced.service; enabled; preset: enabled)
     Active: active (running) since Fri 2024-10-04 16:07:08 UTC; 38s ago
   Main PID: 1466 (nvidia-persiste)
      Tasks: 1 (limit: 18904)
     Memory: 620.0K
        CPU: 15ms
     CGroup: /system.slice/nvidia-persistenced.service
             └─1466 /usr/libexec/nvidia/tesla/bin/nvidia-persistenced --user nvidia --verbose

Oct 04 16:07:06 ip-10-194-20-105.us-west-2.compute.internal systemd[1]: Starting NVIDIA Persistence Daemon...
Oct 04 16:07:06 ip-10-194-20-105.us-west-2.compute.internal nvidia-persistenced[1466]: Verbose syslog connection opened
Oct 04 16:07:06 ip-10-194-20-105.us-west-2.compute.internal nvidia-persistenced[1466]: Now running with user ID 981 and group ID 981
Oct 04 16:07:06 ip-10-194-20-105.us-west-2.compute.internal nvidia-persistenced[1466]: Started (1466)
Oct 04 16:07:08 ip-10-194-20-105.us-west-2.compute.internal nvidia-persistenced[1466]: device 0000:00:1e.0 - registered
Oct 04 16:07:08 ip-10-194-20-105.us-west-2.compute.internal nvidia-persistenced[1466]: device 0000:00:1e.0 - persistence mode enabled.
Oct 04 16:07:08 ip-10-194-20-105.us-west-2.compute.internal nvidia-persistenced[1466]: device 0000:00:1e.0 - NUMA memory onlined.
Oct 04 16:07:08 ip-10-194-20-105.us-west-2.compute.internal nvidia-persistenced[1466]: Local RPC services initialized
Oct 04 16:07:08 ip-10-194-20-105.us-west-2.compute.internal systemd[1]: Started NVIDIA Persistence Daemon.
bash-5.1#
```

bcressey commented on packages/kmod-6.1-nvidia/nvidia-persistenced.service:
```diff
@@ -0,0 +1,15 @@
+ExecStart=/usr/libexec/nvidia/tesla/bin/nvidia-persistenced --user nvidia --verbose
+ExecStopPost=/bin/rm -rf /var/run/nvidia-persistenced
+
+[Install]
+RequiredBy=preconfigured.target
```
If this doesn't specifically need to run in an early phase of boot, I'd just put it with the Fabric Manager in `multi-user.target`:
```suggestion
[Install]
WantedBy=multi-user.target
```

isaac-400 replied:
It can run at any time, but we should prefer that it runs earlier rather than later. Downstream customers may not always initialize the GPU device files themselves, so running this unit early ensures that those files are properly set by the time their units begin (as part of `multi-user.target` for example).

arnaldo2792 replied:
This should run before services like `ecs-gpu-init` so that we can rely on this service to create the devices.

bcressey commented on packages/kmod-6.1-nvidia/nvidia-persistenced.service:
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=NVIDIA Persistence Daemon
+After=load-tesla-kernel-modules.service load-open-gpu-kernel-modules.service
```
If you run in `multi-user.target` instead, you won't need these dependencies.

bcressey commented on packages/kmod-6.1-nvidia/nvidia-persistenced.service:
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=NVIDIA Persistence Daemon
+After=load-tesla-kernel-modules.service load-open-gpu-kernel-modules.service
+# The "nvidia" user must be created before this unit runs.
+Requires=systemd-sysusers.service
+After=systemd-sysusers.service
```
Since you're using the default unit dependencies (i.e. you don't have `DefaultDependencies=no`, which is correct) then you can omit these - they're guaranteed to run before `sysinit.target` is reached.

```suggestion
```

bcressey commented on packages/kmod-6.1-nvidia/nvidia-persistenced.service:
```diff
@@ -0,0 +1,15 @@
+Type=forking
+ExecStartPre=/bin/rm -rf /run/nvidia-persistenced
+ExecStart=/usr/libexec/nvidia/tesla/bin/nvidia-persistenced --user nvidia --verbose
+ExecStopPost=/bin/rm -rf /var/run/nvidia-persistenced
```
Better to use the same paths in both commands
```suggestion
ExecStopPost=/bin/rm -rf /run/nvidia-persistenced
```

bcressey commented on packages/kmod-6.1-nvidia/nvidia-sysusers.conf:
```diff
@@ -0,0 +1 @@
+u nvidia - "nvidia-persistenced user"
```
nit: since we might use it for something else because the username isn't `persistenced` specific:
```suggestion
u nvidia - "nvidia user"
```

bcressey reviewed (commented):
Looks good! Some feedback on the systemd unit, but no blockers.

isaac-400 commented:
Re-built with the `tmpfiles.d` change:

```
● nvidia-persistenced.service - NVIDIA Persistence Daemon
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvidia-persistenced.service; enabled; preset: enabled)
     Active: active (running) since Fri 2024-10-04 20:23:15 UTC; 13min ago
   Main PID: 1436 (nvidia-persiste)
      Tasks: 1 (limit: 18904)
     Memory: 37.1M
        CPU: 1.959s
     CGroup: /system.slice/nvidia-persistenced.service
             └─1436 /usr/libexec/nvidia/tesla/bin/nvidia-persistenced --user nvidia --verbose

Oct 04 20:23:10 ip-10-194-20-136.us-west-2.compute.internal systemd[1]: Starting NVIDIA Persistence Daemon...
Oct 04 20:23:10 ip-10-194-20-136.us-west-2.compute.internal nvidia-persistenced[1436]: Verbose syslog connection opened
Oct 04 20:23:10 ip-10-194-20-136.us-west-2.compute.internal nvidia-persistenced[1436]: Directory /var/run/nvidia-persistenced will not be removed on exit
Oct 04 20:23:10 ip-10-194-20-136.us-west-2.compute.internal nvidia-persistenced[1436]: Now running with user ID 981 and group ID 981
Oct 04 20:23:10 ip-10-194-20-136.us-west-2.compute.internal nvidia-persistenced[1436]: Started (1436)
Oct 04 20:23:10 ip-10-194-20-136.us-west-2.compute.internal nvidia-persistenced[1436]: device 0000:00:1e.0 - registered
Oct 04 20:23:15 ip-10-194-20-136.us-west-2.compute.internal nvidia-persistenced[1436]: device 0000:00:1e.0 - persistence mode enabled.
Oct 04 20:23:15 ip-10-194-20-136.us-west-2.compute.internal nvidia-persistenced[1436]: device 0000:00:1e.0 - NUMA memory onlined.
Oct 04 20:23:15 ip-10-194-20-136.us-west-2.compute.internal nvidia-persistenced[1436]: Local RPC services initialized
Oct 04 20:23:15 ip-10-194-20-136.us-west-2.compute.internal systemd[1]: Started NVIDIA Persistence Daemon.
```

yeazelm reviewed (approved):
LGTM!

Commit:
ecs-gpu-init: run after nvidia-persistenced.service

isaac-400 commented:
Now `ecs-gpu-init` runs after `nvidia-persistenced`:
```
[root@admin]# sudo sheltie
bash-5.1# systemctl status nvidia-persistenced
● nvidia-persistenced.service - NVIDIA Persistence Daemon
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvidia-persistenced.service; enabled; preset: enabled)
     Active: active (running) since Tue 2024-10-08 00:40:40 UTC; 1min 34s ago
   Main PID: 1450 (nvidia-persiste)
      Tasks: 1 (limit: 18904)
     Memory: 37.1M
        CPU: 1.974s
     CGroup: /system.slice/nvidia-persistenced.service
             └─1450 /usr/libexec/nvidia/tesla/bin/nvidia-persistenced --user nvidia --verbose

Oct 08 00:40:36 ip-10-194-20-228.us-west-2.compute.internal systemd[1]: Starting NVIDIA Persistence Daemon...
Oct 08 00:40:36 ip-10-194-20-228.us-west-2.compute.internal nvidia-persistenced[1450]: Verbose syslog connection opened
Oct 08 00:40:36 ip-10-194-20-228.us-west-2.compute.internal nvidia-persistenced[1450]: Directory /var/run/nvidia-persistenced will not be removed on exit
Oct 08 00:40:36 ip-10-194-20-228.us-west-2.compute.internal nvidia-persistenced[1450]: Now running with user ID 981 and group ID 981
Oct 08 00:40:36 ip-10-194-20-228.us-west-2.compute.internal nvidia-persistenced[1450]: Started (1450)
Oct 08 00:40:36 ip-10-194-20-228.us-west-2.compute.internal nvidia-persistenced[1450]: device 0000:00:1e.0 - registered
Oct 08 00:40:40 ip-10-194-20-228.us-west-2.compute.internal nvidia-persistenced[1450]: device 0000:00:1e.0 - persistence mode enabled.
Oct 08 00:40:40 ip-10-194-20-228.us-west-2.compute.internal nvidia-persistenced[1450]: device 0000:00:1e.0 - NUMA memory onlined.
Oct 08 00:40:40 ip-10-194-20-228.us-west-2.compute.internal nvidia-persistenced[1450]: Local RPC services initialized
Oct 08 00:40:40 ip-10-194-20-228.us-west-2.compute.internal systemd[1]: Started NVIDIA Persistence Daemon.
bash-5.1# systemctl status ecs-gpu-init
● ecs-gpu-init.service - Initialize ECS GPU config
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/ecs-gpu-init.service; enabled; preset: enabled)
     Active: active (exited) since Tue 2024-10-08 00:40:40 UTC; 1min 40s ago
   Main PID: 1466 (code=exited, status=0/SUCCESS)
        CPU: 17ms

Oct 08 00:40:40 ip-10-194-20-228.us-west-2.compute.internal systemd[1]: Starting Initialize ECS GPU config...
Oct 08 00:40:40 ip-10-194-20-228.us-west-2.compute.internal systemd[1]: Finished Initialize ECS GPU config.
bash-5.1#
```
